staging/src/k8s.io/apiserver/pkg/endpoints/filters/fail_with_audit.go

```diff
@@ -0,0 +1,100 @@
+/*
+Copyright 2020 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package filters
+
+import (
+	"errors"
+	"fmt"
+	"net/http"
+
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+	auditinternal "k8s.io/apiserver/pkg/apis/audit"
+	"k8s.io/apiserver/pkg/audit"
+	"k8s.io/apiserver/pkg/audit/policy"
+	"k8s.io/apiserver/pkg/endpoints/handlers/responsewriters"
+	"k8s.io/apiserver/pkg/endpoints/request"
+)
+
+// AuditWrapperFunc returns an http.Handler that audits a failed request.
+// The filters that appear before the audit filter in the handler chain can use it
+// to audit a failed request.
+// The specified errorHandler is invoked to render an error response to the request.
+// The specified stringer adapter provides an error message that is to be set in the
+// ResponseStatus field of the audit event.
+type AuditWrapperFunc func(errorHandler http.Handler, stringer func() string) http.Handler
+
+func (w AuditWrapperFunc) Wrap(handler http.Handler, stringer func() string) http.Handler {
```

Comment:
this looks a lot like a filter. But it isn't a filter?

Reply:
@sttts it's not a filter, but it returns a filter that you can use to audit a failed request, this is where I use it - https://github.com/kubernetes/kubernetes/pull/96901/files#diff-e17eb0bdfdb382ac39f2193d8471774402bbd5cbe3de16c7ea986ac301ea5e15R64-R73

```
func withRequestDeadline(handler http.Handler, auditWrapper AuditWrapperFunc, longRunning request.LongRunningRequestCheck,
	negotiatedSerializer runtime.NegotiatedSerializer, requestTimeoutMaximum time.Duration, clock utilclock.PassiveClock) http.Handler {
	return http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {

                if err != nil {    // we need to send a 400 and audit the failed request     
			statusErr := apierrors.NewBadRequest(fmt.Sprintf("%s", err.Error()))

			klog.Errorf("Error - %s: %#v", err.Error(), req.RequestURI)

			failed := failedErrorHandler(negotiatedSerializer, statusErr)
			withAudit := auditWrapper.Wrap(failed, func() string {
				return statusErr.Error()
			})

			withAudit.ServeHTTP(w, req)
              }
```
`statusErr` is request scoped and so I need to build the `fail with audit filter` inside the `withRequestDeadline` filter. Because I am not pushing the  `statusErr` in to the context I can't build the `fail with audit filter` filter inside `DefaultBuildHandlerChain`. Instead I am building a wrapper or factory in `DefaultBuildHandlerChain` and passing the wrapper to  `withRequestDeadline`.

It should have a better name, I agree. does `FailWithAuditBuilderFunc` sound better?